kotlinx-coroutines-core/common/src/flow/operators/Chunk.kt

```diff
@@ -0,0 +1,100 @@
+ * and applies the given [transform] function to an each.
+ *
+ * Note that the list passed to the [transform] function is ephemeral and is valid only inside that function.
+ * You should not store it or allow it to escape in some way, unless you made a snapshot of it.
+ * The last list may have less elements than the given [size].
+ *
+ * This is more efficient, than using flow.chunked(n).map { ... }
+ *
+ * @param size the number of elements to take in each list, must be positive and can be greater than the number of elements in this flow.
+ */
+
+@FlowPreview
+public fun <T, R> Flow<T>.chunked(size: Int, transform: suspend (List<T>) -> R): Flow<R> {
+    require(size > 0) { "Size should be greater than 0, but was $size" }
+    return windowed(size, size, true, transform)
+}
+
+/**
+ * Returns a flow of snapshots of the window of the given [size]
+ * sliding along this flow with the given [step], where each
+ * snapshot is a list.
+ *
+ * Several last lists may have less elements than the given [size].
+ *
+ * Both [size] and [step] must be positive and can be greater than the number of elements in this flow.
+ * @param size the number of elements to take in each window
+ * @param step the number of elements to move the window forward by on an each step
+ * @param partialWindows controls whether or not to keep partial windows in the end if any.
+ */
+
+@FlowPreview
+public fun <T> Flow<T>.windowed(size: Int, step: Int, partialWindows: Boolean): Flow<List<T>> =
+        windowed(size, step, partialWindows) { it.toList() }
+
+/**
+ * Returns a flow of results of applying the given [transform] function to
+ * an each list representing a view over the window of the given [size]
+ * sliding along this collection with the given [step].
+ *
+ * Note that the list passed to the [transform] function is ephemeral and is valid only inside that function.
+ * You should not store it or allow it to escape in some way, unless you made a snapshot of it.
+ * Several last lists may have less elements than the given [size].
+ *
+ * This is more efficient, than using flow.windowed(...).map { ... }
+ *
+ * Both [size] and [step] must be positive and can be greater than the number of elements in this collection.
+ * @param size the number of elements to take in each window
+ * @param step the number of elements to move the window forward by on an each step.
+ * @param partialWindows controls whether or not to keep partial windows in the end if any.
+ */
+
+@OptIn(ExperimentalStdlibApi::class)
+@FlowPreview
+public fun <T, R> Flow<T>.windowed(size: Int, step: Int, partialWindows: Boolean, transform: suspend (List<T>) -> R): Flow<R> {
+    require(size > 0 && step > 0) { "Size and step should be greater than 0, but was size: $size, step: $step" }
+
+    return flow {
+        val buffer = ArrayDeque<T>(size)
+        val toDrop = min(step, size)
+        val toSkip = max(step - size, 0)
+        var skipped = toSkip
+
+        collect { value ->
```

Comment:
Quick observation. If the upstream flow throws an exception, you could end up with a partial buffer here. 
what are your thoughts of adding this before the collect
```
catch { e ->
  if(partialWindows){
    emit(transform(buffer))
  }
  throw e
}.collect { value -> 
```

Reply:
I have fear, that we can loose exception this way or otherwise obscure it.

What if emitting partial buffer gets cancelled on some suspension point, before we rethrow exception? Or even, what if exception will be thrown from emitting our partial buffer, before we manage to rethrow original exception? Perhaps we could fix this by using a finally block somewhere. But, I think, it would be simpler, to just let exception propagate right away, and loose unfortunate partial buffer.

Reply:
Loosing partial buffer is somewhat analogous to coroutineScope cancelling all its children immediately, on detecting exception anywhere. It is just more consistent this way.